rfc/ipni-w3c.md

```diff
@@ -0,0 +1,346 @@
+  - This exposes our block-to-car indexes. Anyone could use IPNI to find which CAR a block is in.
+  -   The context id bytes provide the CAR CID for any block look up. The CAR CID can then be used to find the CAR index via our content-claims API.
+  - We could delete the IPNI records by CAR CID if the CAR is deleted.
+- Make IPNI advertising an explicit UCAN capability that clients can invoke rather than a side-effect of bucket events
+  - With this we are free to write CARs anywhere. The user's agent invokes a `ipni/offer` capability to ask us to publish an IPNI advertisement for the blocks in their CAR.
+  - This empowers the user to opt-in or out as they need, and allows us to bill for the (small) cost of running that service.
+- Provide a way to retrieve a CAR index, that has (CID, offset, length) tuples, without retrieving the CAR file.
+    - Eliminates the need to rely on the CARPark database. Block location information can now be stored on the w3s network.
+    - Clients can request specific data item and only receive data for that item.
+- Put the source inclusion claim in the IPNI advert metadata.
+  - We have to sign IPNI Adverts as the provider. Providing a signed source claim allows more nuanced reputation decisions.
+
+Question: Is there a specific use case for the inclusion claim in the metadata?
+
+### Quick IPNI primer
+
+IPNI ingests and replicates billions of signed provider claims for where individual block CIDs can be retrieved from.
+
+Users can query IPNI servers for any CID, and it provides a set of provider addresses and transport info, along with a provider specific `ContextID` and optional metadata.
+
+For example: <http://cid.contact> hosts an IPNI server that Protocol Labs maintains.
+
+#### Query IPNI for a cid
+ 
+```bash
+curl https://cid.contact/cid/bafybeicawc3qwtlecld6lmtvsndimoz3446xyaprgsxvhd3aapwa2twnc4 -sS | jq
+```
+
+```json
+{
+  "MultihashResults": [
+    {
+      "Multihash": "EiBAsLcLTWQSx+WydZNGhjs75z18AfE0r1OPYAPsDU7NFw==",
+      "ProviderResults": [
+        {
+          "ContextID": "YmFndXFlZXJheTJ2ZWJsZGNhY2JjM3Z0em94bXBvM2NiYmFsNzV3d3R0aHRyamhuaDdvN2o2c2J0d2xmcQ==",
+          "Metadata": "gBI=",
+          "Provider": {
+            "ID": "QmQzqxhK82kAmKvARFZSkUVS6fo9sySaiogAnx5EnZ6ZmC",
+            "Addrs": [
+              "/dns4/elastic.dag.house/tcp/443/wss"
+            ]
+          }
+        },
+        {
+          "ContextID": "YmFndXFlZXJheTJ2ZWJsZGNhY2JjM3Z0em94bXBvM2NiYmFsNzV3d3R0aHRyamhuaDdvN2o2c2J0d2xmcQ==",
+          "Metadata": "oBIA",
+          "Provider": {
+            "ID": "QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp",
+            "Addrs": [
+              "/dns4/dag.w3s.link/tcp/443/https"
+            ]
+          }
+        }
+]}]}
+```
+
+web3.storage publishes the blocks it can provide by encoding a batch of multihashes as an IPLD object and writing it to a bucket as an `Advertisement`, addressed by it's CID.
```

Comment:
I am familiar with a lot of this but it might be helpful to offer a synchronous "brown bag lunch" type learning on how IPNI works, the design goals, etc. In particular, why things like ContextIDs are critical for scale is useful. I'm not 100% sure how much this fully explains what IPNI is intending to do, purely from text.

Reply:
I can certainly do that. This document may provider some explanation:
https://github.com/ipni/storetheindex-docs/blob/main/filecoin-network-indexer/technical-walkthrough.md